Simplify yb-ctl

bmatican opened the pull request:
A few yb-ctl related changes:
- switched error logging to a temp file and added cleanup mechanics: on success, remove the log file; on error, dump it to stdout with a note that we encountered errors
- also added a `--verbose` flag which would always dump logging to stderr
- make `yb-ctl` work from the `yugabyte-installation` repo as well, with a locally built `yugabyte-db`, assuming both repos are cloned into the same top-level directory, i.e. `<top_level_dir>/yugabyte-db` (or `<top_level_dir>/yugabyte`), and `<top_level_dir>/yugabyte-installation`
- change the status format output, both for the cluster and for each individual node
- add correct waiting for the cluster to come up on both create/start/restart/...
- change default number of drives to 1 and expose a `--num_drives` option
- persist replication factor to config file
- fix function for checking if ysql is enabled or not
- add `timeout` param (and remove unused params) and timeout-respecting wrapper function
- minor: change some internal variable names and external string outputs from redis/cql/postgres to yedis/ycql/ysql

Sample output (also added a custom `--data_dir` just for examples):

Sample `create`:
```
./bin/yb-ctl --data_dir /tmp/yugabyte-local-cluster create
----------------------------------------------------------------------------------------------------
| Number of servers: 1 | Replication factor: 1                                                     |
----------------------------------------------------------------------------------------------------
| JDBC                : postgresql://postgres@127.0.0.1:5433                                       |
| YSQL                : ./bin/psql -U postgres -h 127.0.0.1 -p 5433                                |
| YCQL                : ./bin/cqlsh 127.0.0.1 9042                                                 |
| YEDIS               : ./bin/redis-cli -h 127.0.0.1 -p 6379                                       |
| Web UI              : http://127.0.0.1:7000/                                                     |
| Cluster Data        : /tmp/yugabyte-local-cluster                                                |
----------------------------------------------------------------------------------------------------

For more info, please use: yb-ctl --data_dir /tmp/yugabyte-local-cluster status
```

Sample `add_node`:
```
./bin/yb-ctl --data_dir /tmp/yugabyte-local-cluster add_node
----------------------------------------------------------------------------------------------------
| Node 2: yb-tserver (7830)                                                                        |
----------------------------------------------------------------------------------------------------
| JDBC                : postgresql://postgres@127.0.0.2:5433                                       |
| YSQL                : ./bin/psql -U postgres -h 127.0.0.2 -p 5433                                |
| YCQL                : ./bin/cqlsh 127.0.0.2 9042                                                 |
| YEDIS               : ./bin/redis-cli -h 127.0.0.2 -p 6379                                       |
| data-dir[0]         : /tmp/yugabyte-local-cluster/node-2/disk-1/yb-data                          |
| TServer Logs        : /tmp/yugabyte-local-cluster/node-2/disk-1/yb-data/tserver/logs             |
----------------------------------------------------------------------------------------------------
```

Sample `status` after `create` + `add_node` + `stop_node 1` (note the `Stopped` vs PID in headers):
```
 ./bin/yb-ctl --data_dir /tmp/yugabyte-local-cluster status
----------------------------------------------------------------------------------------------------
| Cluster UUID: 4c33a667-a085-4182-a27c-be94462afef0                                               |
----------------------------------------------------------------------------------------------------
| Master Count        : 1                                                                          |
| TServer Count       : 2                                                                          |
| Replication Factor  : 1                                                                          |
| Web UI              : http://127.0.0.1:7000/                                                     |
| Cluster Data        : /tmp/yugabyte-local-cluster                                                |
----------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------
| Node 1: yb-tserver (Stopped), yb-master (7395)                                                   |
----------------------------------------------------------------------------------------------------
| JDBC                : postgresql://postgres@127.0.0.1:5433                                       |
| YSQL                : ./bin/psql -U postgres -h 127.0.0.1 -p 5433                                |
| YCQL                : ./bin/cqlsh 127.0.0.1 9042                                                 |
| YEDIS               : ./bin/redis-cli -h 127.0.0.1 -p 6379                                       |
| data-dir[0]         : /tmp/yugabyte-local-cluster/node-1/disk-1/yb-data                          |
| TServer Logs        : /tmp/yugabyte-local-cluster/node-1/disk-1/yb-data/tserver/logs             |
| Master Logs         : /tmp/yugabyte-local-cluster/node-1/disk-1/yb-data/master/logs              |
----------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------
| Node 2: yb-tserver (7830)                                                                        |
----------------------------------------------------------------------------------------------------
| JDBC                : postgresql://postgres@127.0.0.2:5433                                       |
| YSQL                : ./bin/psql -U postgres -h 127.0.0.2 -p 5433                                |
| YCQL                : ./bin/cqlsh 127.0.0.2 9042                                                 |
| YEDIS               : ./bin/redis-cli -h 127.0.0.2 -p 6379                                       |
| data-dir[0]         : /tmp/yugabyte-local-cluster/node-2/disk-1/yb-data                          |
| TServer Logs        : /tmp/yugabyte-local-cluster/node-2/disk-1/yb-data/tserver/logs             |
----------------------------------------------------------------------------------------------------
```

Sample error by running two `create` calls back to back:
```
# NON VERBOSE
./bin/yb-ctl --data_dir /tmp/yugabyte-local-cluster create  --disable_ysql
2019-04-18 21:54:14,108 ERROR: Found cluster data at /tmp/yugabyte-local-cluster, cannot start new cluster. Use --data_dir to specify a different data directory if necessary, or 'destroy' / 'wipe_restart'.Commands to wipe out the old directory.
^^^ Encountered errors ^^^

# VERBOSE
./bin/yb-ctl --data_dir /tmp/yugabyte-local-cluster --verbose create --disable_ysql
2019-04-18 21:54:20,727 ERROR: Found cluster data at /tmp/yugabyte-local-cluster, cannot start new cluster. Use --data_dir to specify a different data directory if necessary, or 'destroy' / 'wipe_restart'.Commands to wipe out the old directory.
```

bmatican committed:
Simplify yb-ctl

bmatican committed:
Remove debug changes

bmatican committed:
Fix lint

mbautin commented on bin/yb-ctl:
```diff
@@ -497,6 +507,11 @@ class ClusterOptions:
+        num_drives = int(getattr(args, "num_drives", 1))
+        if num_drives <= 0:
+            raise ExitWithError("Invalid number of drives: {}".format(num_drives))
+        self.drives = ["disk-{}".format(i) for i in xrange(1, num_drives + 1)]
```
It would be good to move away from xrange to be one step closer to Python 3 compatibility.

ndeodhar replied:
nit: xrange has been removed in python 3. So, it may be better to use range to make it easier to upgrade in future.

mbautin commented on bin/yb-ctl:
```diff
@@ -1054,16 +1074,31 @@ class ClusterControl:
+                if idx == 0:
+                    info_list.append(("log path", os.path.join(data_path, "logs")))
+                # TODO: do we want this??
+                # info_list.append(("data path[{}]".format(idx), data_path))
```
Let's uncomment or remove.

mbautin commented on bin/yb-ctl:
```diff
@@ -1260,8 +1303,11 @@ class ClusterControl:
             level=logging.INFO,
             format="%(asctime)s %(levelname)s: %(message)s")
+    print("Logging debug information to: {}".format(
+        os.path.join(os.getcwd(), SCRIPT_DEBUG_FILE)))
```
I am not so sure about creating an odd `startup.log` file in the current directory. Also the output of every invocation of `yb-ctl` will now be cluttered with the "Logging debug information" message. Perhaps let's disable debug logging by default and still print it to stderr if a `--verbose` (`-v`) flag is specified?

bmatican replied:
are you suggesting not logging anything anywhere if not `verbose`? just doing prints to the console for things like status?

mbautin replied:
Yes, that's an option. In my opinion, command-line tools for interacting with servers should not produce a log file in the current directory unless explicitly asked to.

bmatican replied:
I'm still not sure which way to go on this. I definitely do not want this to not log anything by default! It will clearly fail in various weird ways and if there's no log neither users nor us will be able to know what happened...

However, I also agree that dumping a log file in wherever we called the script from, is not ideal...The best workaround for that I think would be to drop it into the cluster data dir, but there's at least 2 issues with that that I can think of:
- we have to carefully massage when we setup when we parse the args (so we know what dir we need) vs when we setup the logging (so we can actually respect logging.info calls afterwards) vs when we do the rest of the code, including the one that actually creates the directory in which we want to log...
- if users run commands with and without specifying the data_dir, we'll end up sprawling some log files in various locations, potentially even causing inadvertent failures (say we create the default dir and put our log file there, then they stop using a custom dir and try the usual one, but it already exists, with only this log file there..)

bmatican replied:
Discussed offline, added mechanics to dump to a temp file, clean it up if we did not have errors, else dump it to stderr if we did have errors. Also added `--verbose` to always dump logging to stderr regardless of success/fail

ndeodhar commented on bin/yb-ctl:
```diff
@@ -1054,16 +1074,31 @@ class ClusterControl:
                 info_list.extend([
-                    ("cql service", get_address("cql_rpc")),
-                    ("redis service", get_address("redis_rpc")),
+                    ("ycql service", "cqlsh {} {}".format(
```
It's better to print out the full path (./bin/cqlsh) instead of just binary name (cqlsh).

ndeodhar commented on bin/yb-ctl:
```diff
@@ -1054,16 +1074,31 @@ class ClusterControl:
-                    info_list.append(("pgsql service", get_address("pgsql_rpc")))
-            logging.info("Server is running: {}".format(
-                ", ".join(["{}={}".format(k, v) for k, v in info_list])))
+                    ysql_port = self.options.base_ports[daemon_id.daemon_type]["ysql_rpc"]
```
Since we'll have more SQL users, we should print this before YCQL and YEDIS.

ndeodhar commented on bin/yb-ctl:
```diff
@@ -1001,12 +1020,11 @@ class ClusterControl:
         num_yb_admin_ts = None
         start_time = time.time()
+        logging.info("Waiting for master leader election tablet server registration.")
         while wait_count < MAX_WAIT_ITERS and (time.time() - start_time <= MAX_WAIT_SECONDS):
```
Will be good to add an argument to increase timeout (similar to yb-docker-ctl).

ndeodhar reviewed (commented):
Awesome Bogdan!

bmatican committed:
Addressing review comments. Changes include:
- adding correct handling for replication_factor
- adding backwards compatible handling for num_drives
- fixing handling of is_postgres_enabled due to None vs false
- removed unused --require_clock_sync
- added os.devnull stderr pipe for some subprocess calls
- changed test.sh to include some add_node and stop_node calls and use non-standard data_dir, to bypass default NFS for some home setups

Also changed status reporting to better reflect new direction:
- `create` will report a cluster section
- `status` will report a cluster section and a list of node sections
- `add_node`, `start_node` will report a node section for that id
- cluster section will have server counts, RF and the web UI
- node sections have API connection info, logs, PIDs and proccess status

bmatican committed:
Some more changes:
- fixed redis-cli type
- made cluster UI link point to master port 7000
- add timeout feature
- refactor the two timeout loops to use a timeout-aware wrapper

bmatican committed:
Remove startup.log and replace with only bubbling up errors on actual failures

ndeodhar commented on bin/yb-ctl:
```diff
@@ -88,10 +90,25 @@ DAEMON_TYPES = [
+        timeout_sec = MAX_WAIT_SECONDS
+    start_time = time.time()
+    wait_count = 0
+    while wait_count < MAX_WAIT_ITERS and (time.time() - start_time <= timeout_sec):
```
I think we should remove MAX_WAIT_ITERS and just use timeout_sec. Using MAX_WAIT_ITERS will end up not honoring timeout_sec. For example, if user's setup is slow and she increases timeout to X, it's likely that we'll timeout even before we reach X seconds due to MAX_WAIT_ITERS.

mbautin replied:
+1

ndeodhar commented on bin/yb-ctl:
```diff
@@ -1022,48 +1072,120 @@ class ClusterControl:
+            # TODO: this might cause issues if the first master is down...
+            ("Web UI", "http://{}/".format(
+                self.options.get_address(DaemonId(DAEMON_TYPE_MASTER, 1), "http", True))),
+            ("Cluster Data", self.options.cluster_base_dir)
```
I think we should show YSQL, JDBC etc. info for 1 node on cluster creation as well. It doesn't present a complete picture of our system otherwise.

bmatican replied:
Honestly, I would even consider printing the whole status (so cluster + all node blocks), but I think Karthik felt that would be too cluttered...I'm not sure if printing just 1 node would be ideal though, as that might be confusing if we have a mode where we print all the nodes and at startup we just print 1?

cc @rkarthik007 @schoudhury

ndeodhar replied:
I didn't mean print details for one node. What I meant is print YSQL, JDBC etc. with the IP of first node. That way, users know how to connect to our DB. For a common single node install, it's anyway the only way to connect. So, output of create will be something like this:
```
./bin/yb-ctl --data_dir /tmp/yugabyte-local-cluster create
----------------------------------------------------------------------------------------------------
| Cluster UUID: 4c33a667-a085-4182-a27c-be94462afef0                                               |
----------------------------------------------------------------------------------------------------
| Master Count        : 1                                                                          |
| TServer Count       : 1                                                                          |
| Replication Factor  : 1                                                                          |
| Web UI              : http://127.0.0.1:7000/                                                     |
| Cluster Data        : /tmp/yugabyte-local-cluster                                                |
| JDBC                : postgresql://postgres@127.0.0.1:5433                                       |
| YSQL                : ./bin/psql -U postgres -h 127.0.0.1 -p 5433                                |
| YCQL                : ./bin/cqlsh 127.0.0.1 9042                                                 |
| YEDIS               : ./bin/redis-cli -h 127.0.0.1 -p 6379                                       |
----------------------------------------------------------------------------------------------------
```

mbautin commented on bin/yb-ctl:
```diff
@@ -88,10 +90,25 @@ DAEMON_TYPES = [
 
+def retry_fn_with_timeout(fn, timeout_sec=None):
+    if timeout_sec is None:
+        timeout_sec = MAX_WAIT_SECONDS
```
Why not assign the default value in function definition directly?

mbautin commented on bin/yb-ctl:
```diff
@@ -88,10 +90,25 @@ DAEMON_TYPES = [
+    wait_count = 0
+    while wait_count < MAX_WAIT_ITERS and (time.time() - start_time <= timeout_sec):
+        if fn():
+            return True
```
Should we return the return value of `fn()` here? It is also obviously "truthy" in this case, but might have extra meaning to the caller.

mbautin commented on bin/yb-ctl:
```diff
@@ -88,10 +90,25 @@ DAEMON_TYPES = [
+            return True
+        wait_count += 1
+        time.sleep(SLEEP_TIME_IN_SEC)
+    return False
```
Remove this line, to effectively return `None`.

mbautin commented on bin/yb-ctl:
```diff
@@ -403,6 +421,9 @@ class ClusterOptions:
 
         self.installation_dir = None
         yb_src_root_candidate = os.path.dirname(os.path.dirname(os.path.dirname(self.script_dir)))
+        # For internal development, assume $PARENT_DIR/yugabyte-installation/bin/yb-ctl will have
```
internal development -> YugaByte DB development (this workflow will also apply to all contributors to YugaByte DB).

mbautin reviewed (commented):
A few more comments, need to review some more.

bmatican committed:
Addressed most of the comments and wired modify_placement_info into the new retry wrapper model

bmatican committed:
Adding connectivity info to the cluster status box and message for using the status command for more info

bmatican committed:
Fix lint

mbautin commented on bin/yb-ctl:
```diff
@@ -497,6 +514,10 @@ class ClusterOptions:
                                    "separated by commas.")
             self.placement_info.append(t_item)
 
+        self.num_drives = getattr(args, "num_drives", None)
```
Do we really need "None" here as the third argument? I think `args.num_drives` will always exist, but it will sometimes be None and that's OK (it won't raise an AttributeError).

bmatican replied:
Yeah, it throws `AttributeError: 'Namespace' object has no attribute 'num_drives'` otherwise for commands that do not have `num_drives` setup, which i was trying to avoid.

mbautin commented on bin/yb-ctl:
```diff
@@ -616,8 +632,24 @@ class ClusterControl:
+        return self.cluster_config.get(
+            "enable_postgres",
+            self.cluster_config.get(
+                "enable_ysql", self.options.enable_ysql))
```
If neither enable_postgres nor enable_ysql are explicitly set to false, let's consider YSQL enabled.

bmatican replied:
I think the logic here is that, if we have a line in the config, we'll use it (so any new deployments, AFTER the create), otherwise (old deployments or just creating), we use the arg value.

mbautin commented:
Still need to finish another round of code review.

mbautin commented on bin/yb-ctl:
```diff
@@ -414,22 +430,22 @@ class ClusterOptions:
                 yb_src_root_candidate,  'build', 'latest'),
-            os.path.join(yb_src_root_candidate + '__build', 'latest')
+            os.path.join(yb_src_root_candidate + '__build', 'latest'),
+            # yb-ctl being run from a sibling repo to "yugabyte" or "yugabyte-db".
```
Mention here that this is a development-only case as well.

mbautin commented on bin/yb-ctl:
```diff
@@ -616,8 +632,24 @@ class ClusterControl:
     def is_postgres_enabled(self):
-        return (self.cluster_config.get("enable_postgres") or
-                self.cluster_config.get("enable_ysql") or self.options.enable_ysql)
+        return self.cluster_config.get(
```
Let's also rename the function to is_ysql_enabled.

bmatican replied:
Replaced this and a couple more that felt ok to change

mbautin commented on bin/yb-ctl:
```diff
@@ -616,8 +632,24 @@ class ClusterControl:
+        return self.cluster_config.get("replication_factor") or self.options.replication_factor
+
+    def get_num_drives(self):
+        # If the config does not have an entry, use the old default.
```
We could figure out the number of drives from the directory hierarchy for existing clusters.

mbautin commented on bin/yb-ctl:
```diff
@@ -726,30 +765,39 @@ class ClusterControl:
+                logging.info("Successfully modified placement info.")
+                return True
+            except subprocess.CalledProcessError:
+                pass
```
We should capture stdout/stderr/error code from the exception and expose it in the error message in case of failure.

bmatican replied:
I think the problem is that an error is expected in these types of calls if the cluster is not currently up or master leader is not up, etc..that's why we're retry looping over these...I guess one option could be to augment the `retry_fn_with_timeout` to be smart about when the last possible call of the wrapped function is, and pass in a bool for whether or not to capture this info and log it? thoughts @mbautin

mbautin commented on bin/yb-ctl:
```diff
@@ -902,11 +950,11 @@ class ClusterControl:
         last_msg_time = time.time()
 
-        def print_wait_msg():
+        def wait_log_msg():
```
log_waiting_msg?

mbautin commented on bin/yb-ctl:
```diff
@@ -977,20 +1025,22 @@ class ClusterControl:
             except subprocess.CalledProcessError:
-                wait_count += 1
-                time.sleep(SLEEP_TIME_IN_SEC)
+                pass
```
See my earlier comment regarding silently ignoring errors.

mbautin commented on bin/yb-ctl:
```diff
@@ -1022,48 +1070,135 @@ class ClusterControl:
+                else:
+                    logging.info("Waiting for all tablet servers to register: {}/{}".format(
+                        num_yb_admin_ts, num_alive_ts))
             except subprocess.CalledProcessError:
```
Ditto regarding suppressing errors.

mbautin commented on bin/yb-ctl:
```diff
@@ -1022,48 +1070,135 @@ class ClusterControl:
+            # We will display None if the subprocess fails...
+            cluster_uuid = regex_groups[0] if regex_groups else None
+        except subprocess.CalledProcessError as e:
+            pass
```
Ditto regarding suppressing errors.

mbautin commented on bin/yb-ctl:
```diff
@@ -1022,48 +1070,135 @@ class ClusterControl:
+        try:
+            with open(os.devnull, "wb") as devnull:
+                cluster_uuid = subprocess.check_output(
+                    "curl localhost:7000/cluster-config | grep cluster_uuid",
```
Let's not add a dependency on curl -- let's send the HTTP call in Python.

bmatican replied:
I just removed this all together, since we decided not to do the `cluster_uuid` at all

mbautin commented on bin/yb-ctl:
```diff
@@ -1022,48 +1070,135 @@ class ClusterControl:
+
+    def cluster_status(self, more_info=True):
+        # cluster_uuid = self.query_cluster_uuid()
+        # title = "Cluster UUID: {}".format(cluster_uuid)
```
Remove commented-out code.

bmatican committed:
Address comments

bmatican committed:
Adding retry changes to log out errors on the last attempt

bmatican committed:
Fix PID check

mbautin commented on bin/yb-ctl:
```diff
@@ -85,11 +87,27 @@ DAEMON_TYPES = [
+
 
-YB_POSTGRES_PORT = 5433
+def retry_fn_with_timeout(fn, timeout_sec=MAX_WAIT_SECONDS):
```
Perhaps this should be renamed to `retry_call_with_timeout` now that it has specific processing for `CalledProcessError`.

mbautin commented on bin/yb-ctl:
```diff
@@ -85,11 +87,27 @@ DAEMON_TYPES = [
+    while True:
+        time_elapsed = time.time() - start_time
+        if time_elapsed > timeout_sec:
+            break
```
This is a bit counter-intuitive for a function like this: in case of timeout there is no exception, but the function just returns. This needs to be documented in the docstring.

mbautin commented on bin/yb-ctl:
```diff
@@ -717,6 +762,10 @@ class ClusterControl:
                                          help="Disable YugaByte SQL API.")
+
+            subparsers[cmd].add_argument("--num_drives", default=1, type=int,
+                                         help="Number of drives to emulate.")
```
Perhaps expand the description to mention that we create separate directories and pretend they are separate drives.

mbautin commented on bin/yb-ctl:
```diff
@@ -726,30 +775,39 @@ class ClusterControl:
+
+        def call_modify_placement_info():
+            with open(os.devnull, "wb") as devnull:
+                subprocess.check_call(cmd, stderr=devnull)
```
We won't get any error output even if it fails all the way up until the timeout.

mbautin commented on bin/yb-ctl:
```diff
@@ -726,30 +775,39 @@ class ClusterControl:
+                return 0
             return len(glob.glob("{}/*/{}/yb-data/{}".format(
-                self.options.cluster_base_dir, self.options.drives[0], daemon_type)))
+                self.options.cluster_base_dir, drives[0], daemon_type)))
```
Are we assuming that the list of node ids represented by these directories is contiguous? Or should we perhaps extract the max node id from the directory names?

bmatican replied:
We are indeed assuming this and it is supposed to be how we allocate them when the user adds new nodes.

mbautin commented on bin/yb-ctl:
```diff
@@ -977,20 +1036,19 @@ class ClusterControl:
+
+        def call_change_config():
+            with open(os.devnull, "wb") as devnull:
+                subprocess.check_output(cmd, stderr=devnull)
```
ditto regarding redirecting stderr to /dev/null

mbautin commented on bin/yb-ctl:
```diff
@@ -1000,70 +1058,135 @@ class ClusterControl:
+            if proc.returncode:
+                raise subprocess.CalledProcessError(proc.returncode,
+                                                    cmd_list_tservers, output=output)
+            num_yb_admin_ts = len(output.splitlines()) - 1
```
Perhaps some validation on the output would be useful.

mbautin replied:
I mean, not just assuming a certain output format and counting the lines, but maybe searching for lines in a certain format, and showing the output to the user in case it is invalid .

mbautin commented on bin/yb-ctl:
```diff
@@ -977,20 +1036,19 @@ class ClusterControl:
-        # should be equal to the replication factor.
-        wait_count = 0
+        # 'Enough' here means that the number of live tservers reported to the master should be
+        # equal to the number of TS we have started -- based on directories we have on the FS.
```
Document the return value.

mbautin commented on bin/yb-ctl:
```diff
@@ -1000,70 +1058,135 @@ class ClusterControl:
+        info_kv_list.extend([
+            # TODO: this might cause issues if the first master is down...
+            ("Web UI", "http://{}/".format(
+                self.options.get_address(DaemonId(DAEMON_TYPE_MASTER, 1), "http", True))),
```
perhaps use `k=v` syntax for `True`

mbautin reviewed (commented):
Still need to finish review.

mbautin commented on bin/yb-ctl:
```diff
@@ -1000,70 +1058,135 @@ class ClusterControl:
+        logging.info("Waiting for master leader election tablet server registration.")
+        if not retry_fn_with_timeout(call_check_for_ts, self.options.timeout):
+            logging.error("Failed waiting for {} tservers, got {}".format(
+                num_alive_ts, num_yb_admin_ts))
```
Are the num_alive_ts and num_yb_admin_ts up-to-date enough at this point?

bmatican replied:
On success, yes. On failure, it is updated with whatever actually got valid values...

mbautin commented on bin/yb-ctl:
```diff
@@ -1257,14 +1410,28 @@ class ClusterControl:
+            # If we called the atexit handler and we have a file, it means we must have had errors.
+            if keep_log_file_and_dump_to_stderr:
+                with open(self.log_file, "rb") as log:
+                    shutil.copyfileobj(log, sys.stdout)
```
Shouldn't we print to stderr here instead of stdout?

mbautin commented on test/test.sh:
```diff
@@ -9,8 +9,21 @@ log() {
 
+start_cluster_run_tests() {
+  root_dir=$1
+  $root_dir/yb-ctl $flag_data_dir start
```
double quotes around `$root_dir`

mbautin commented on test/test.sh:
```diff
@@ -9,8 +9,21 @@ log() {
 yb_data_dirs=(
-  "$HOME/yugabyte-data"
+  "/tmp/yugabyte-data"
+)
```
Not sure why we need an array here if we're only using one data dir anyway.

bmatican replied:
I assumed your intention when you set this up was that we might add multiple data dirs

bmatican committed:
Addressing review comments.
